tsdb/index/tsi1/index.go

```diff
@@ -590,7 +607,54 @@ func (i *Index) CreateSeriesListIfNotExists(keys [][]byte, names [][]byte, tagsS
 				if idx >= len(i.partitions) {
 					return // No more work.
 				}
-				errC <- i.partitions[idx].createSeriesListIfNotExists(pNames[idx], pTags[idx])
+
+				ids, err := i.partitions[idx].createSeriesListIfNotExists(pNames[idx], pTags[idx])
+
+				var updateCache bool
+				for _, id := range ids {
+					if id != 0 {
+						updateCache = true
+						break
+					}
+				}
+
+				if !updateCache {
+					errC <- err
+					continue
+				}
+
+				// Some cached bitset results may need to be updated.
+				i.tagValueCache.mu.RLock()
+				for j, id := range ids {
+					if id == 0 {
+						continue
+					}
+
+					name := pNames[idx][j]
+					tags := pTags[idx][j]
+					if i.tagValueCache.measurementContainsSets(name) {
+						for _, pair := range tags {
+							// TODO(edd): It's not clear to me yet whether it will be better to take a lock
+							// on every series id set, or whether to gather them all up under the cache rlock
+							// and then take the cache lock and update them all at once (without invoking a lock
+							// on each series id set).
+							//
+							// Taking the cache lock will block all queries, but is one lock. Taking each series set
+							// lock might be many lock/unlocks but will only block a query that needs that particular set.
+							//
+							// Need to think on it, but I think taking a lock on each series id set is the way to go.
+							//
+							// One other option here is to take a lock on the series id set when we first encounter it
+							// and then keep it locked until we're done with all the ids.
+							//
+							// Note: this will only add `id` to the set if it exists.
+							i.tagValueCache.addToSet(name, pair.Key, pair.Value, id) // Takes a lock on the series id set
```

Comment:
one other option to consider here is to just drop the cache entries for the `name`. that will be less good if measurements have a really big scope, though.

Reply:
Adding to the set under lock is tens of nanoseconds, depending on the value being added. Rebuilding the cached series id set is hundreds of microseconds. I think if there was constant series creation for a general tag (like "region = west") then dropping the cached set each time would be more expensive.